File: finlib-cpp/src/finlib/portfolio/portfolio_asset.cpp
//
// Created by Andy Pack on 19/06/2025.
//

#include "portfolio_asset.hpp"

#include <vector>

namespace finlib {
    PortfolioAsset::PortfolioAsset(std::string name, double quantity, finlibrs::TimeSpan timespan) {
        handle = finlibrs::portfolio_asset_new(reinterpret_cast<uint8_t *>(name.data()),
                                               static_cast<int32_t>(name.length()),
                                               quantity,
                                               timespan);
    }

    PortfolioAsset::PortfolioAsset(finlibrs::PortfolioAsset *handle) {
        this->handle = handle;
    }

    // double PortfolioAsset::current_value() {
    //     return finlibrs::portfolio_asset_current_value(handle);
    // }

    finlibrs::NullableFloat PortfolioAsset::current_total_value() {
        return finlibrs::portfolio_asset_current_total_value(handle);
    }

    finlibrs::NullableFloat PortfolioAsset::profit_loss() {
        return finlibrs::portfolio_asset_profit_loss(handle);
    }

    finlibrs::Tuple PortfolioAsset::get_mean_and_std() {
        return finlibrs::portfolio_asset_get_mean_and_std(handle);
    }

    PortfolioAsset::~PortfolioAsset() {
        if (handle)
            finlibrs::portfolio_asset_destroy(handle);
    }
}

File: FinLib.NET/FinLib/Portfolio/PortfolioAsset.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using FinLib.Extensions;
using FinLib.Price;
using FinLib.Risk;

namespace FinLib.Portfolio;

public class PortfolioAsset: IDisposable, IPayoff<double?>, IProfit<double?>, IValueAtRisk
{
    private readonly unsafe PortfolioAsset_native* _handle;
    internal unsafe PortfolioAsset_native* GetPtr() => _handle;

    public PortfolioAsset(string assetName, double quantity, TimeSpan timeSpan)
    {
        unsafe
        {
            var n = Encoding.UTF8.GetBytes(assetName);
            fixed (byte* namePtr = n){
                _handle = NativeMethods.portfolio_asset_new(
                    namePtr, assetName.Length, quantity, timeSpan.MapTimeSpan());
            }
        }
    }

     public PriceRangePair? CurrentValue
     {
         get
         {
             unsafe
             {
                 var item = NativeMethods.portfolio_asset_current_value(_handle);
                 if (item is not null)
                 {
                     return new PriceRangePair(item);
                 }

                 return null;
             }
         }
     }

    public double? CurrentTotalValue
    {
        get
        {
            unsafe
            {
                return NativeMethods.portfolio_asset_current_total_value(_handle);
            }
        }
    }

    public double? ProfitLoss
    {
        get
        {
            unsafe
            {
                return NativeMethods.portfolio_asset_profit_loss(_handle);
            }
        }
    }

    public double Payoff(double? underlying = null)
    {
        unsafe
        {
            return NativeMethods.portfolio_asset_payoff(_handle, underlying);
        }
    }

    public double Profit(double? underlying = null)
    {
        unsafe
        {
            return NativeMethods.portfolio_asset_profit(_handle, underlying);
        }
    }

    public double? ValueAtRisk(double confidence, double? initialInvestment = null)
    {
        unsafe
        {
            return NativeMethods.portfolio_asset_value_at_risk(_handle, confidence, initialInvestment);
        }
    }

    public double? ValueAtRiskPercent(double confidence)
    {
        unsafe
        {
            return NativeMethods.portfolio_asset_value_at_risk_percent(_handle, confidence);
        }
    }

    public double? ValueAtRiskAfterTime(double confidence, nint at, double? initialInvestment = null)
    {
        unsafe
        {
            return NativeMethods.portfolio_asset_value_at_risk_afer_time(_handle, confidence, initialInvestment, at);
        }
    }

    private void ReleaseUnmanagedResources()
    {
        unsafe
        {
            NativeMethods.portfolio_asset_destroy(_handle);
        }
    }

    public void Dispose()
    {
        ReleaseUnmanagedResources();
        GC.SuppressFinalize(this);
    }

    ~PortfolioAsset()
    {
        ReleaseUnmanagedResources();
    }
}